### ISSUE TITLE
[crash] memory allocation failed error during parsing

### ISSUE THREAD

**Issue by pventuzelo** [rhysd/wain#30]
Hey,

Just triggered this memory allocation bug. Don't know if your cargo-fuzz was able to catch it so I prefer to open an issue anyway ;) 

```
$ ./target/debug/wain alloc_failed_wain.wasm
memory allocation of 23477014352 bytes failed[1]    10077 abort (core dumped)  ./target/debug/wain alloc_failed_wain.wasm

```

Download: [alloc_failed_wain.zip](https://github.com/rhysd/wain/files/4770644/alloc_failed_wain.zip)

# backtrace from gdb

```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff71e1801 in __GI_abort () at abort.c:79
#2  0x0000555555711a37 in std::sys::unix::abort_internal () at src/libstd/sys/unix/mod.rs:167
#3  0x000055555570f62d in rust_oom () at src/libstd/alloc.rs:282
#4  0x000055555572a9a7 in alloc::alloc::handle_alloc_error () at src/liballoc/alloc.rs:307
#5  0x00005555556a273c in alloc::raw_vec::RawVec<T,A>::reserve (self=0x7fffffffac38, used_capacity=0, 
    needed_extra_capacity=266784254)
    at /rustc/49cae55760da0a43428eba73abcb659bb70cf2e4/src/liballoc/raw_vec.rs:345
#6  0x00005555556b19a9 in alloc::vec::Vec<T>::reserve (self=0x7fffffffac38, additional=266784254)
    at /rustc/49cae55760da0a43428eba73abcb659bb70cf2e4/src/liballoc/vec.rs:500
#7  0x00005555556a52cd in <wain_ast::Module as wain_syntax_binary::parser::Parse>::parse (
    parser=0x7fffffffd650) at wain-syntax-binary/src/parser.rs:325
#8  0x0000555555683827 in wain_syntax_binary::parser::Parser::parse (self=0x7fffffffd650)
    at wain-syntax-binary/src/parser.rs:126
#9  0x00005555556a2be2 in <wain_ast::Root<wain_syntax_binary::source::BinarySource> as wain_syntax_binary::parser::Parse>::parse (parser=0x7fffffffd650) at wain-syntax-binary/src/parser.rs:218
#10 0x0000555555683707 in wain_syntax_binary::parser::Parser::parse (self=0x7fffffffd650)
    at wain-syntax-binary/src/parser.rs:126
#11 0x00005555556806c8 in wain_syntax_binary::parse (input=...) at wain-syntax-binary/src/lib.rs:16
#12 0x00005555555930cb in wain::run_binary (bin=...) at src/main.rs:150
#13 0x00005555555933f1 in wain::main () at src/main.rs:179
```

**Comment by rhysd** [comment]
Yes, I also triggered the same issue with `cargo fuzz`. And running `cargo run` with the same input did not trigger the OOM crash.

```
> cargo run alloc_failed_wain.wasm
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s
     Running `target/debug/wain alloc_failed_wain.wasm`
Error on parsing: expected one of 0x7f, 0x7e, 0x7d, 0x7c for value type but got byte 0x01 while parsing global section. caused at byte offset 14

 ... 01 7f 03 01 80 80 9f 0e 00 00 01 00 4f 4f 4f 4f 4f 0f fd 4c 54 28 14 35 30
     ^
     starts from here
```

This is because `cargo fuzz` runs fuzzer with RSS limit `2048Mb`. I'm not sure that this is actually a bug.

**Comment by pventuzelo** [comment]
using `cargo run` I trigger the crash as well:
```
$ cargo run alloc_failed_wain.wasm
    Finished dev [unoptimized + debuginfo] target(s) in 0.03s
     Running `target/debug/wain alloc_failed_wain.wasm`
memory allocation of 23477014352 bytes failed[1]    26931 abort (core dumped)  cargo run alloc_failed_wain.wasm
```

**Comment by pventuzelo** [comment]
wasm-validate detect that the module is invalid
```
$ wasm-validate alloc_failed_wain.wasm
000000e: error: invalid global count 266784254, only 24 bytes left in section
```

I think your parser allocate the globals before checking the section validity.

**Comment by rhysd** [comment]
Here is a summary I got from `cargo fuzz`

```
==14933== ERROR: libFuzzer: out-of-memory (malloc(2583691176))
   To change the out-of-memory limit use -rss_limit_mb=<N>

    #0 0x108465045 in __sanitizer_print_stack_trace+0x35 (librustc-nightly_rt.asan.dylib:x86_64h+0x4f045)
    #1 0x107ccd3e2 in fuzzer::PrintStackTrace() FuzzerUtil.cpp:205
    #2 0x107cab3ee in fuzzer::Fuzzer::HandleMalloc(unsigned long) FuzzerLoop.cpp:130
    #3 0x107cab233 in fuzzer::MallocHook(void const volatile*, unsigned long) FuzzerLoop.cpp:99
    #4 0x10846b5ba in __sanitizer::RunMallocHooks(void const*, unsigned long)+0x1a (librustc-nightly_rt.asan.dylib:x86_64h+0x555ba)
    #5 0x108419a46 in __asan::Allocator::Allocate(unsigned long, unsigned long, __sanitizer::BufferedStackTrace*, __asan::AllocType, bool)+0x816 (librustc-nightly_rt.asan.dylib:x86_64h+0x3a46)
    #6 0x108419208 in __asan::asan_malloc(unsigned long, __sanitizer::BufferedStackTrace*)+0x28 (librustc-nightly_rt.asan.dylib:x86_64h+0x3208)
    #7 0x10845bd4b in wrap_malloc+0xfb (librustc-nightly_rt.asan.dylib:x86_64h+0x45d4b)
    #8 0x107c8eb36 in alloc::alloc::alloc::hd266daa3740076ac alloc.rs:80
    #9 0x107c90085 in _$LT$alloc..alloc..Global$u20$as$u20$core..alloc..AllocRef$GT$::alloc::h011d981bf79805d2 alloc.rs:174
    #10 0x107c91dfe in alloc::raw_vec::finish_grow::h1680bcb9f7d39ad0 raw_vec.rs:514
    #11 0x107bb3ed1 in alloc::raw_vec::RawVec$LT$T$C$A$GT$::grow_amortized::h649b87ff493bc594 raw_vec.rs:440
    #12 0x107bad995 in alloc::raw_vec::RawVec$LT$T$C$A$GT$::try_reserve::h6504414e47357017 raw_vec.rs:281
    #13 0x107bbd5a3 in alloc::raw_vec::RawVec$LT$T$C$A$GT$::reserve::hfb92161d76ef410f raw_vec.rs:267
    #14 0x107c6df8b in alloc::vec::Vec$LT$T$GT$::reserve::h5065619728a4bf2e vec.rs:500
    #15 0x107c02b89 in _$LT$wain_ast..Module$u20$as$u20$wain_syntax_binary..parser..Parse$GT$::parse::h69b5dcf285f57ac4 parser.rs:310
    #16 0x107bd4fe6 in wain_syntax_binary::parser::Parser::parse::ha025a033f2f0d22a parser.rs:126
    #17 0x107bfbe3b in _$LT$wain_ast..Root$LT$wain_syntax_binary..source..BinarySource$GT$$u20$as$u20$wain_syntax_binary..parser..Parse$GT$::parse::he287170c56bf22a1 parser.rs:218
    #18 0x107bd5026 in wain_syntax_binary::parser::Parser::parse::hbc5ce223ae2fb081 parser.rs:126
    #19 0x107bc2971 in wain_syntax_binary::parse::h42ea5e43561b7c12 lib.rs:16
    #20 0x107b945e8 in rust_fuzzer_test_input binary_parser.rs:5
    #21 0x107ca5a62 in libfuzzer_sys::test_input_wrap::_$u7b$$u7b$closure$u7d$$u7d$::h2629f014b9239b24 lib.rs:27
    #22 0x107ca6c63 in std::panicking::try::do_call::h0b729c537da6eba6 panicking.rs:297
    #23 0x107ca7120 in __rust_try+0x30 (binary_parser:x86_64+0x100117120)
    #24 0x107ca67d6 in std::panicking::try::hc708128672bd0eca panicking.rs:274
    #25 0x107ca9d5e in std::panic::catch_unwind::h7be88b52f7253974 panic.rs:394
    #26 0x107ca53be in LLVMFuzzerTestOneInput lib.rs:25
    #27 0x107cae01d in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) FuzzerLoop.cpp:553
    #28 0x107d0bee1 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) FuzzerDriver.cpp:292
    #29 0x107d113fd in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) FuzzerDriver.cpp:775
    #30 0x107d3d5c9 in main FuzzerMain.cpp:19
    #31 0x7fff60dfd3d4 in start+0x0 (libdyld.dylib:x86_64+0x163d4)

SUMMARY: libFuzzer: out-of-memory
```

Though the position OOM occurred is a bit different from yours, this has the same root cause. Let me explain.

Wasm binary format defines `vec(B)` structure. It is something like variadic length.

https://webassembly.github.io/spec/core/binary/conventions.html#vectors

It contains the length of vector at first `u32` value. And the same number of elements are followed.

`wain_syntax_binary::Parser` at first parses the first `u32` value and reserves memory on `Vec` with `Vec::reserve` (e.g. [here](https://github.com/rhysd/wain/blob/5cbf87c9536c2df1db5ec22c6bbaa18435218bdb/wain-syntax-binary/src/parser.rs#L325)). However, when the binary incorrectly sets too large value to the first `u32` value, the `Vec::reserve` method crashes. To avoid this, we need [`Vec::try_reserve`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.try_reserve). But it is not stabilized yet so we need to wait for stabilization.

**Comment by rhysd** [comment]
> I think your parser allocate the globals before checking the section validity.

OK, I think I can add check by looking length of input before `Vec::reserve`. This is not perfect because input may contain the number of elements actually (imagine more than 4GB wasm binary). But almost all cases can be detected by checking input length in advance.

**Comment by rhysd** [comment]
I fixed this at 6b09f8c.

After the fix, the fuzzer could run for 10 minutes without crash.
